DRA: migrate DeviceTaint and Conditions to declarative validation

dumko2001 opened the pull request:
What type of PR is this?
/kind cleanup
/kind feature

What this PR does / why we need it:
This PR migrates the validation for Device.Taints and AllocatedDeviceStatus.Conditions from manual Go code to declarative markers, addressing items from the tracking issue #135073.

Changes:

staging/src/k8s.io/api/resource/v1/types.go:
Added // +k8s:maxItems=16 to Device.Taints.
Added // +k8s:maxItems=8, // +k8s:listType=map, and // +k8s:listMapKey=type to AllocatedDeviceStatus.Conditions.
pkg/apis/resource/validation/validation.go:
Updated validateDevice to mark Taints size as covered by declarative validation (sizeCovered).
Removed the manual length check for Conditions in validateDeviceStatus.
Ran make update to synchronize generated files (zz_generated.* and generated.proto).

Which issue(s) this PR is related to:
Fixes #135073

Special notes for your reviewer:
The tracking issue suggested adding // +k8s:unique=map for AllocatedDeviceStatus.Conditions.

I have omitted this specific tag because it causes a conflict in update-codegen:

tag "k8s:unique": unique tag may not be used with listType=set or listType=map
Since listType=map already enforces uniqueness on the key (type), the explicit unique tag is redundant.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

lalitc375 commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -1282,9 +1282,6 @@ func validateDeviceStatus(device resource.AllocatedDeviceStatus, fldPath *field.
 		allErrs = append(allErrs, field.Invalid(fldPath, sharedDeviceID, "must be an allocated device in the claim"))
 	}
-	if len(device.Conditions) > resource.AllocatedDeviceStatusMaxConditions {
-		allErrs = append(allErrs, field.TooMany(fldPath.Child("conditions"), len(device.Conditions), resource.AllocatedDeviceStatusMaxConditions))
```
Hand written validation should not be removed. it should be marked by `.MarkCoveredByDeclarative()` Look other references.

dumko2001 replied:
done

lalitc375 reviewed (commented):
Thank you for the PR @dumko2001 . Totally appreciate it.

Please follow this pattern for adding tags.

```
The ideal PR structure for a PR contributing a DV tag migration is:

Commit # 1: IF NOT ALREADY WIRED UP FOR DV - Wire up the apigroup, add empty test and empty generated file. I don't need to review this, you know the pattern by now.

Commit # 2: Use a single DV tag on a single field. E.g. Add +k8s:required to required field. Includes generated code and robust test cases. Emphasis on test-cases -- we want enough to prove that a field is correct, but we don't want to duplicate all the per-tag tests (e.g. a k8s-short-name field has dozens of corner cases, but we really only need a few). We want each case to be as short and specific as possible, which is why the "tweak" pattern and the use of Origin for Invalid() errors is so important. Specify as little as possible to be sufficiently precise. Newcomers need guidance on this.

Commits # 3...N: Use DV tags on the same field, one per commit; repeat until that field is done or all that is left is not handled by DV yet. Includes generated code and tests for each field. (not the case for the specific examples below as they are for single tag migrations)

Commits N+1...M: Repeat the above for more fields. (not the case for the specific examples below as they are for single field migrations)

The emphasis is on lots of small commits. Maintaining commit discipline is hard but worthwhile.

If DV exposes bugs in the existing testing or the hand-written validation itself (!!), those should be fixed as a PR of its own or as commits at the front of the PR.
```

lalitc375 commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -1283,9 +1283,16 @@ func validateDeviceStatus(device resource.AllocatedDeviceStatus, fldPath *field.
+		allErrs = append(allErrs, field.TooMany(fldPath.Child("conditions"), len(device.Conditions), resource.AllocatedDeviceStatusMaxConditions).WithOrigin("maxItems").MarkCoveredByDeclarative())
+	}
+	// Allow declarative validation to handle duplicates (via +listType=map)
+	for _, err := range metav1validation.ValidateConditions(device.Conditions, fldPath.Child("conditions")) {
```
You should add options in the ValidationConditions method to mark which validations are covered by DV. Look into ValidateCSIDriverName method for example for reference.

https://github.com/kubernetes/kubernetes/blob/b2ac9e206fdd912f35f2ab5b3c5b5243303ba14b/pkg/apis/core/validation/validation.go#L1817

lalitc375 commented on staging/src/k8s.io/api/resource/v1/types.go:
```diff
@@ -650,6 +651,7 @@ type DeviceTaint struct {
 	// Must be a label name.
 	//
 	// +required
+	// +k8s:validation:MinLength=1
```
The tag `+k8s:validation:MinLength` is **not a valid tag** for core API types. The `validation:` namespace is used by controller-gen for CRDs but is not recognized by the Kubernetes core API machinery. Please remove this tag or replace it with a supported one if the machinery allows (e.g., `+k8s:minLength=1` if supported, otherwise keep it handwritten).

lalitc375 reviewed (commented):
The migration of handwritten validation to declarative validation for DRA is mostly correct for the `Conditions` field. However, there are some issues to address:

1. **Invalid Tags**: The `+k8s:validation:*` tags are not supported in core API types.
2. **Missing +k8s:optional**: `AllocatedDeviceStatus.Conditions` is missing the declarative `+k8s:optional` tag.
3. **Handwritten Coverage**: The `ValidateLabelName` call for `taint.Key` in `validation.go` should remain handwritten and not be marked as covered until a valid declarative tag is implemented.
4. **Test Regression**: Duplicate condition type testing was removed and not re-added.

lalitc375 commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -1283,9 +1283,16 @@ func validateDeviceStatus(device resource.AllocatedDeviceStatus, fldPath *field.
+		allErrs = append(allErrs, field.TooMany(fldPath.Child("conditions"), len(device.Conditions), resource.AllocatedDeviceStatusMaxConditions).WithOrigin("maxItems").MarkCoveredByDeclarative())
+	}
+	// Allow declarative validation to handle duplicates (via +listType=map)
+	for _, err := range metav1validation.ValidateConditions(device.Conditions, fldPath.Child("conditions")) {
```
You should add options in the ValidationConditions method to mark which validations are covered by DV. Look into ValidateCSIDriverName method for example for reference.

https://github.com/kubernetes/kubernetes/blob/b2ac9e206fdd912f35f2ab5b3c5b5243303ba14b/pkg/apis/core/validation/validation.go#L1817

dumko2001 replied:
Done. Duplicate errors marked as covered by DV.

dumko2001 committed:
api: migrate Device.Taints to declarative validation

dumko2001 committed:
Fix DRA validation: logic, tags, tests, and cleanup

dumko2001 committed:
fix: address review feedback for DRA declarative validation

- Remove invalid +k8s:validation:MinLength=1 tag from DeviceTaint.Key
- Remove redundant +k8s:validation:Enum tag from DeviceTaint.Effect
- Add missing +k8s:optional tag to AllocatedDeviceStatus.Conditions
- Applied to v1, v1beta1, and v1beta2 API versions

dumko2001 committed:
api: migrate AllocatedDeviceStatus.Conditions to declarative size validation

dumko2001 committed:
api: migrate AllocatedDeviceStatus.Conditions to declarative map validation

dumko2001 commented on staging/src/k8s.io/api/resource/v1/types.go:
```diff
@@ -371,6 +371,7 @@ type Device struct {
 	// +optional
 	// +listType=atomic
```
done

dumko2001 commented on staging/src/k8s.io/api/resource/v1/types.go:
```diff
@@ -371,6 +371,7 @@ type Device struct {
 	// +optional
 	// +listType=atomic
 	// +featureGate=DRADeviceTaints
+	// +k8s:maxItems=16
```
done

lalitc375 commented on pkg/registry/resource/resourceclaim/declarative_validation_test.go:
```diff
@@ -1738,3 +1739,76 @@ func tweakStatusDevicesTooManyIPs(count int) func(rc *resource.ResourceClaim) {
+					Driver: "test", Pool: "test", Device: "test",
+					Conditions: validConditions,
+				}))
+				claim.ObjectMeta.ResourceVersion = "1"
```
This can move to the test.

lalitc375 commented on pkg/registry/resource/resourceclaim/declarative_validation_test.go:
```diff
@@ -1738,3 +1739,76 @@ func tweakStatusDevicesTooManyIPs(count int) func(rc *resource.ResourceClaim) {
+		old          resource.ResourceClaim
+		expectedErrs field.ErrorList
+	}{
+		"invalid: too many conditions": {
```
Also add an atlimit test case.

dumko2001 committed:
test: add listMapKey, at-limit tests for DRA declarative validation

- Add 'valid: at limit taints' test (16 taints) for ResourceSlice
- Add 'valid: at limit conditions' test (8 conditions) for ResourceClaim
- Add 'invalid: duplicate condition type' test for +k8s:listMapKey=type
- Move ResourceVersion assignment to test loop
- Add normalization rule for conditions path matching
- Mark duplicate condition errors as covered by declarative

dumko2001 commented on pkg/registry/resource/resourceslice/declarative_validation_test.go:
```diff
@@ -91,6 +91,12 @@ func TestDeclarativeValidate(t *testing.T) {
 							resource.DeviceTaintEffect("Invalid"), []string{}),
 					},
 				},
+				"invalid: max taints exceeded": {
```
Done

dumko2001 commented:
@lalitc375 Addressed all 4 comments:

Added +k8s:listMapKey=type test with normalization rule
Added at-limit taints test (16 taints)
Moved ResourceVersion to test loop
Added at-limit conditions test (8 conditions)

lalitc375 commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -64,6 +64,13 @@ var ResourceNormalizationRules = []field.NormalizationRule{
+		// handwritten ValidateConditions errors (conditions[N].type)
+		// This is applied to declarative errors (got) to match handwritten format (want)
+		Regexp:      regexp.MustCompile(`\.conditions\[(\d+)\]$`),
+		Replacement: ".conditions[$1].type",
```
Remove .type from handwritten code. We have decided to not use the key name to show duplicate errors or any errors related to the list constraints.  Fieldpaths in errors can be changed,as they are not considered part of the API.

lalitc375 commented on pkg/registry/resource/resourceclaim/declarative_validation_test.go:
```diff
@@ -1738,3 +1739,120 @@ func tweakStatusDevicesTooManyIPs(count int) func(rc *resource.ResourceClaim) {
+	}
+}
+
+func testDeclarativeValidateStatus(t *testing.T, apiVersion string) {
```
Do we need creation tests for status?

dumko2001 replied:
No ,status only has update path, no create operation.

dumko2001 committed:
fix: declarative validation equivalence tests for resource claim status

lalitc375 commented:
Please rebase the PR.

dumko2001 committed:
fix: restore unrelated generated validation files after rebase

dumko2001 committed:
test: fix declarative validation checks after rebase

dumko2001 committed:
fix: align DRA declarative validation after rebase

dumko2001 committed:
test: explicitly enable legacy DRA claim APIs in kind

dumko2001 committed:
test: fix DRA kind kubeadm v1beta4 runtime-config wiring

dumko2001 commented:
@lalitc375 rebased PR is now green on `a7b6b48f36ea`.

The small follow-up in `test/e2e/dra/kind.yaml` was only to keep the existing DRA kind presubmits working after the rebase onto current master: kinds `kubeadm.k8s.io/v1beta4` `apiServer.extraArgs` needed the new list form for `runtime-config`, otherwise the legacy DRA claim/template APIs used by those jobs were not served and the DRA kind jobs failed with 404s.

Could you please take another look when you have a moment?

lalitc375 commented:
I just realized this has overlap with #139101  . Can you remove DV from conditions in this PR? Please also rebase
